src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorDiagnosticsEndpoint.cs

```diff
@@ -0,0 +1,176 @@
+            {
+                // No diagnostics left after filtering.
+                return new RazorDiagnosticsResponse()
+                {
+                    Diagnostics = Array.Empty<Diagnostic>(),
+                    HostDocumentVersion = documentVersion
+                };
+            }
+
+            return await MapDiagnosticsAsync(
+                request,
+                filteredDiagnostics,
+                documentVersion,
+                documentSnapshot).ConfigureAwait(false);
+
+            // TODO; HTML filtering blocked on https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1257401
+            static bool CanDiagnosticBeFiltered(Diagnostic d) =>
+                (DiagnosticsToIgnore.Contains(d.Code?.String) &&
+                 d.Severity != DiagnosticSeverity.Error);
+        }
+
+        private async Task<RazorDiagnosticsResponse> MapDiagnosticsAsync(
+            RazorDiagnosticsParams request,
+            Diagnostic[] filteredDiagnostics,
+            int? documentVersion,
+            DocumentSnapshot documentSnapshot)
+        {
+            if (request.Kind != RazorLanguageKind.CSharp)
+            {
+                // All other non-C# requests map directly to where they are in the document.
+                return new RazorDiagnosticsResponse()
+                {
+                    Diagnostics = filteredDiagnostics,
+                    HostDocumentVersion = documentVersion,
+                };
+            }
+
+            var codeDocument = await documentSnapshot.GetGeneratedOutputAsync().ConfigureAwait(false);
+            var mappedDiagnostics = new List<Diagnostic>();
+
+            for (var i = 0; i < filteredDiagnostics.Length; i++)
+            {
+                var diagnostic = filteredDiagnostics.ElementAt(i);
+                var projectedRange = diagnostic.Range;
+
+                if (codeDocument.IsUnsupported() ||
+                    !_documentMappingService.TryMapFromProjectedDocumentRange(
+                        codeDocument, projectedRange,
+                        MappingBehavior.Inclusive,
+                        out var originalRange))
+                {
+                    // Couldn't remap the range correctly.
+                    // If this isn't an `Error` Severity Diagnostic we can discard it.
+                    if (diagnostic.Severity != DiagnosticSeverity.Error)
+                    {
+                        continue;
+                    }
+
+                    // For `Error` Severity diagnostics we still show the diagnostics to
+                    // the user, however we set the range to an undefined range to ensure
+                    // clicking on the diagnostic doesn't cause errors.
+                    originalRange = RangeExtensions.UndefinedRange;
+                }
+
+                diagnostic.Range = originalRange;
+                mappedDiagnostics.Add(diagnostic);
+            }
+
+            return new RazorDiagnosticsResponse()
```

Comment:
I wouldn't expect MapDiagnosticsAsync to return the response. That's more of the `Handle`'s responsibility. Makes more sense for this to return an `IReadOnlyList<Diagnostic>` and have the consumer construct the response